packages/deployments/contracts/contracts/messaging/connectors/axelar/BaseAxelar.sol

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity 0.8.17;
+
+import {AxelarExecutable} from "@axelar-network/axelar-gmp-sdk-solidity/contracts/executable/AxelarExecutable.sol";
+import {IAxelarGateway} from "@axelar-network/axelar-gmp-sdk-solidity/contracts/interfaces/IAxelarGateway.sol";
+import {IAxelarGasService} from "@axelar-network/axelar-gmp-sdk-solidity/contracts/interfaces/IAxelarGasService.sol";
+import {StringToAddress, AddressToString} from "@axelar-network/axelar-gmp-sdk-solidity/contracts/utils/AddressString.sol";
+
+abstract contract BaseAxelar is AxelarExecutable {
+  using StringToAddress for string;
+  using AddressToString for address;
+
+  // ============ Internal Storage ============
+  IAxelarGasService public immutable gasService;
+
+  // Mirror chain id
+  string public MIRROR_CHAIN_ID;
+
+  // ============ Constructor ============
+  constructor(address _amb, bytes memory _axelarParams) AxelarExecutable(_amb) {
+    (address _gasReceiver, string memory _mirrorChainId) = abi.decode(_axelarParams, (address, string));
+    // sanity checks
+    require(bytes(_mirrorChainId).length != 0, "!mirrorChainId");
+
+    // set immutable propertioes
+    gasService = IAxelarGasService(_gasReceiver);
+    MIRROR_CHAIN_ID = _mirrorChainId;
+  }
+
+  // ============ Public Fns ============
+  function _execute(
+    string calldata _sourceChain,
+    string calldata _sourceAddress,
+    bytes calldata _payload
+  ) internal override {
+    require(keccak256(abi.encodePacked(_sourceChain)) == keccak256(abi.encodePacked(MIRROR_CHAIN_ID)), "!source chain");
+    _processMessageFrom(_sourceAddress.toAddress(), _payload);
+  }
+
+  // ============ Private fns ============
+  // DO NOT override _processMessage, should revert from `Connector` class. All messages must use the _processMessageFrom function
+  /**
+   * @notice This function is called to handle incoming messages. Should store the latest
+   * root generated on the l2 domain.
+   */
+  function _processMessageFrom(address sender, bytes memory _data) internal virtual;
+
+  /**
+   * @dev Sends `outboundRoot` to root manager on the mirror chain
+   */
+  function _sendMessage(address _mirrorConnector, bytes memory _data, bytes memory _encodedData) internal {
+    // Should always be sending a merkle root
+    require(_data.length == 32, "!data length");
+
+    // Should not include any gas info
+    require(_encodedData.length == 0, "!data length");
+
+    string memory mirrorConnectorStr = _mirrorConnector.toString();
+
+    // Get the min fees
+    gasService.payNativeGasForContractCall{value: msg.value}(
+      address(this),
+      MIRROR_CHAIN_ID,
+      mirrorConnectorStr,
+      _data,
+      msg.sender
```

Comment:
the refund address here will be the `RootManager`, which has no way to withdraw funds

Reply:
there's also no way to bump gas here, which may be needed?

also:
- this should only happen if `msg.value > 0`
- its not clear who excess fees should be refunded to to me here

Reply:
We should bum gas fee on their explorer manually, there is no way to do that onchain

Reply:
you [can do it onchain](https://docs.axelar.dev/dev/general-message-passing/gas-services/increase-gas#addnativegas), but we can default to the explorer

Reply:
it will just be more annoying to track down

Reply:
I don't think we will ever be bumping gas because we will just wait for the next root to be sent in the next cycle. So I'm not sure this should be part of the functionality of this contract if we can also use the explorer.

For the refund address it should be passed in, it can be our safe, since we are the ones paying the fees.